Add new Blazor tutorial on BWA with EF Core (#32747)
dotnet/AspNetCore.Docs: aspnetcore/blazor/tutorials/movie-database-app/part-4.md

zone_pivot_groups: tooling
---
# Build a Blazor movie database app (Part 4 - Work with a database)

<!-- UPDATE 9.0 Activate after release

[!INCLUDE[](~/includes/not-latest-version.md)]

-->

This article is the fourth part of the Blazor movie database app tutorial that teaches you the basics of building an ASP.NET Core Blazor Web App with features to manage a movie database.

This part of the tutorial series focuses on the database context and directly working with the database's schema and data. Seeding the database with data is also covered.

## Database context

The database context, `BlazorWebAppMoviesContext`, connects to the database and maps model objects to database records. The database context was created in the second part of this series. The scaffolded database context code appears in the `Program` file:

:::zone pivot="vs"

```csharp
builder.Services.AddDbContext<BlazorWebAppMoviesContext>(options =>
    options.UseSqlServer(
      builder.Configuration.GetConnectionString("BlazorWebAppMoviesContext") ?? 
      throw new InvalidOperationException(
         "Connection string 'BlazorWebAppMoviesContext' not found.")));
```

:::zone-end

:::zone pivot="vsc"

```csharp

> builder.Services.AddDbContext<BlazorWebAppMoviesContext>(options =>

Reviewer: ```suggestion
builder.Services.AddDbContextFactory<BlazorWebAppMoviesContext>(options =>
```

    options.UseSqlite(
      builder.Configuration.GetConnectionString("BlazorWebAppMoviesContext") ?? 
      throw new InvalidOperationException(
         "Connection string 'BlazorWebAppMoviesContext' not found.")));
```

:::zone-end

:::zone pivot="cli"

```csharp

> builder.Services.AddDbContext<BlazorWebAppMoviesContext>(options =>

Reviewer: ```suggestion
builder.Services.AddDbContextFactory<BlazorWebAppMoviesContext>(options =>
```

    options.UseSqlite(
      builder.Configuration.GetConnectionString("BlazorWebAppMoviesContext") ?? 
      throw new InvalidOperationException(
         "Connection string 'BlazorWebAppMoviesContext' not found.")));
```

:::zone-end

> <xref:Microsoft.Extensions.DependencyInjection.EntityFrameworkServiceCollectionExtensions.AddDbContext%2A> registers the given context as a service in the app's service collection.

Reviewer: ```suggestion
<xref:Microsoft.Extensions.DependencyInjection.EntityFrameworkServiceCollectionExtensions.AddDbContextFactory%2A> registers a factory for the given context as a service in the app's service collection.
```

* The <xref:Microsoft.AspNetCore.Components.Forms.EditForm.Model?displayProperty=nameWithType> parameter specifies the top-level model object for the form. An edit context is constructed for the form using the assigned model.
* The form is displayed with the values from the movie.

When the `Edit` page is posted to the server, the form values on the page are bound to the `Movie` property because the [`[SupplyParameterFromForm]` attribute](xref:Microsoft.AspNetCore.Components.SupplyParameterFromFormAttribute) is annotated on the `Movie` property:

```csharp
[SupplyParameterFromForm]
public Movie? Movie { get; set; }
```

If the model state has errors when the form is posted, for example if `ReleaseDate` can't be converted into a date, the form is redisplayed with the submitted values. If no model errors exist, the movie is saved using the form's posted values.

## Concurrency exception handling

Review the `UpdateMovie` method of the `Edit` component (`Components/Pages/MoviePages/Edit.razor`):

```csharp
public async Task UpdateMovie()
{

>     DB.Attach(Movie!).State = EntityState.Modified;
> 
>     try
>     {
>         await DB.SaveChangesAsync();
>     }

Reviewer: ```suggestion
    using var context = DbFactory.CreateDbContext();
    context.Attach(Movie!).State = EntityState.Modified;

    try
    {
        await context.SaveChangesAsync();
    }
```

There are many third-party tools you can use to manage and view SQLite databases. The following image shows [DB Browser for SQLite](https://sqlitebrowser.org/):

![DB Browser for SQLite showing movie database](~/blazor/tutorials/movie-database-app/part-4/_static/dbb.png)

In this tutorial, EF Core migrations are used. A migration updates the database schema to match changes in the data model. However, migrations can only make changes to the database that the EF Core provider supports. Resources are listed at the end of this article for further reading.

:::zone-end

## Seed the database

Seeding code can create a set of records for development testing or even be used to create the initial data for a new production database.

> [!WARNING]
> The technique shown in this section isn't acceptable for modifying a live database with active users.

Create a new class named `SeedData` in the `Data` folder with the following code.

`Data/SeedData.cs`:

```csharp
using Microsoft.EntityFrameworkCore;
using BlazorWebAppMovies.Models;

namespace BlazorWebAppMovies.Data
{
    public class SeedData

>     {
>         public static void Initialize(IServiceProvider serviceProvider)
>         {
>             using var context = new BlazorWebAppMoviesContext(
>                 serviceProvider.GetRequiredService<
>                     DbContextOptions<BlazorWebAppMoviesContext>>());
> 
>             if (context == null || context.Movie == null)
>             {
>                 throw new NullReferenceException(
>                     "Null BlazorWebAppMoviesContext or Movie DbSet");
>             }
> 
>             if (context.Movie.Any())
>             {
>                 return;
>             }
> 
>             context.Movie.AddRange(
>                 new Movie
>                 {
>                     Title = "Mad Max",
>                     ReleaseDate = new DateOnly(1979, 4, 12),
>                     Genre = "Sci-fi (Cyberpunk)",
>                     Price = 2.51M,
>                 },
>                 new Movie
>                 {
>                     Title = "The Road Warrior",
>                     ReleaseDate = new DateOnly(1981, 12, 24),
>                     Genre = "Sci-fi (Cyberpunk)",
>                     Price = 2.78M,
>                 },
>                 new Movie
>                 {
>                     Title = "Mad Max: Beyond Thunderdome",
>                     ReleaseDate = new DateOnly(1985, 7, 10),
>                     Genre = "Sci-fi (Cyberpunk)",
>                     Price = 3.55M,
>                 },
>                 new Movie
>                 {
>                     Title = "Mad Max: Fury Road",
>                     ReleaseDate = new DateOnly(2015, 5, 15),
>                     Genre = "Sci-fi (Cyberpunk)",
>                     Price = 8.43M,
>                 },
>                 new Movie
>                 {
>                     Title = "Furiosa: A Mad Max Saga",
>                     ReleaseDate = new DateOnly(2024, 5, 24),
>                     Genre = "Sci-fi (Cyberpunk)",
>                     Price = 13.49M,
>                 }
>             );
> 
>             context.SaveChanges();
>         }
>     }
> }

Author: ```suggestion
namespace BlazorWebAppMovies.Data;

public class SeedData
{
    public static void Initialize(IServiceProvider serviceProvider)
    {
        using var context = new BlazorWebAppMoviesContext(
            serviceProvider.GetRequiredService<
                DbContextOptions<BlazorWebAppMoviesContext>>());

        if (context == null || context.Movie == null)
        {
            throw new NullReferenceException(
                "Null BlazorWebAppMoviesContext or Movie DbSet");
        }

        if (context.Movie.Any())
        {
            return;
        }

        context.Movie.AddRange(
            new Movie
            {
                Title = "Mad Max",
                ReleaseDate = new DateOnly(1979, 4, 12),
                Genre = "Sci-fi (Cyberpunk)",
                Price = 2.51M,
            },
            new Movie
            {
                Title = "The Road Warrior",
                ReleaseDate = new DateOnly(1981, 12, 24),
                Genre = "Sci-fi (Cyberpunk)",
                Price = 2.78M,
            },
            new Movie
            {
                Title = "Mad Max: Beyond Thunderdome",
                ReleaseDate = new DateOnly(1985, 7, 10),
                Genre = "Sci-fi (Cyberpunk)",
                Price = 3.55M,
            },
            new Movie
            {
                Title = "Mad Max: Fury Road",
                ReleaseDate = new DateOnly(2015, 5, 15),
                Genre = "Sci-fi (Cyberpunk)",
                Price = 8.43M,
            },
            new Movie
            {
                Title = "Furiosa: A Mad Max Saga",
                ReleaseDate = new DateOnly(2024, 5, 24),
                Genre = "Sci-fi (Cyberpunk)",
                Price = 13.49M,
            });

        context.SaveChanges();
    }
}
```

> [!WARNING]
> The technique shown in this section isn't acceptable for modifying a live database with active users.

Create a new class named `SeedData` in the `Data` folder with the following code.

`Data/SeedData.cs`:

```csharp
using Microsoft.EntityFrameworkCore;
using BlazorWebAppMovies.Models;

namespace BlazorWebAppMovies.Data
{
    public class SeedData
    {
        public static void Initialize(IServiceProvider serviceProvider)
        {
            using var context = new BlazorWebAppMoviesContext(
                serviceProvider.GetRequiredService<
                    DbContextOptions<BlazorWebAppMoviesContext>>());

            if (context == null || context.Movie == null)
            {
                throw new NullReferenceException(
                    "Null BlazorWebAppMoviesContext or Movie DbSet");
            }

            if (context.Movie.Any())
            {
                return;
            }

            context.Movie.AddRange(
                new Movie
                {
                    Title = "Mad Max",

>                     ReleaseDate = DateTime.Parse("1979-4-12"),

Author: ```suggestion
                    ReleaseDate = DateOnly.Parse("1979-4-12"),
```

                    Genre = "Sci-fi (Cyberpunk)",
                    Price = 2.51M,
                },
                new Movie
                {
                    Title = "The Road Warrior",

>                     ReleaseDate = DateTime.Parse("1981-12-24"),

Author: ```suggestion
                    ReleaseDate = DateOnly.Parse("1981-12-24"),
```

                    Genre = "Sci-fi (Cyberpunk)",
                    Price = 2.78M,
                },
                new Movie
                {
                    Title = "Mad Max: Beyond Thunderdome",

>                     ReleaseDate = DateTime.Parse("1985-7-10"),

Author: ```suggestion
                    ReleaseDate = DateOnly.Parse("1985-7-10"),
```

                    Genre = "Sci-fi (Cyberpunk)",
                    Price = 3.55M,
                },
                new Movie
                {
                    Title = "Mad Max: Fury Road",

>                     ReleaseDate = DateTime.Parse("2015-5-15"),

Author: ```suggestion
                    ReleaseDate = DateOnly.Parse("2015-5-15"),
```

                    Genre = "Sci-fi (Cyberpunk)",
                    Price = 8.43M,
                },
                new Movie
                {
                    Title = "Furiosa: A Mad Max Saga",

>                     ReleaseDate = DateTime.Parse("2024-5-24"),

Author: ```suggestion
                    ReleaseDate = DateOnly.Parse("2024-5-24"),
```
